### ISSUE TITLE
No ALPN Processor for sun.security.ssl.SSLEngineImpl from [org.eclipse.jetty.alpn.conscrypt.server.ConscryptServerALPNProcessor@2f4a2218]

### ISSUE THREAD

**Issue by CaoManhDat** [eclipse/jetty.project#2950]
Hi guys,

When I tried to run an embedded Jetty instance with this setup 
```java
      final SslContextFactory sslcontext = SSLConfig.createContextFactory(config.sslConfig);

      HttpConfiguration configuration = new HttpConfiguration();
      ServerConnector connector;
      if (sslcontext != null) {
        Security.addProvider(new OpenSSLProvider());
        sslcontext.setProvider("Conscrypt");
        sslcontext.setCipherComparator(HTTP2Cipher.COMPARATOR);
        sslcontext.setUseCipherSuitesOrder(true);

        configuration.setSecureScheme("https");
        configuration.addCustomizer(new SecureRequestCustomizer());
        connector = new ServerConnector(server);
        SslConnectionFactory sslConnectionFactory = new SslConnectionFactory(sslcontext, "alpn");
        connector.addConnectionFactory(sslConnectionFactory);
        connector.setDefaultProtocol(sslConnectionFactory.getProtocol());

        HttpConnectionFactory http1ConnectionFactory = new HttpConnectionFactory(configuration);
        HTTP2ServerConnectionFactory http2ConnectionFactory = new HTTP2ServerConnectionFactory(configuration);

        ALPNServerConnectionFactory alpn = new ALPNServerConnectionFactory(
            http2ConnectionFactory.getProtocol(),
            http1ConnectionFactory.getProtocol());
        alpn.setDefaultProtocol(http1ConnectionFactory.getProtocol());
        connector.addConnectionFactory(alpn);
        connector.addConnectionFactory(http1ConnectionFactory);
        connector.addConnectionFactory(http2ConnectionFactory);
      }
```

I met this exception when clients try to connect to that Jetty instance.
```
47668 DEBUG (qtp764074479-22) [    ] o.e.j.i.ManagedSelector 
java.lang.IllegalStateException: Connection rejected: No ALPN Processor for sun.security.ssl.SSLEngineImpl from [org.eclipse.jetty.alpn.conscrypt.server.ConscryptServerALPNProcessor@2f4a2218]
	at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.newServerConnection(ALPNServerConnectionFactory.java:110) ~[jetty-alpn-server-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.server.NegotiatingServerConnectionFactory.newConnection(NegotiatingServerConnectionFactory.java:109) ~[jetty-server-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.server.SslConnectionFactory.newConnection(SslConnectionFactory.java:95) ~[jetty-server-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.server.ServerConnector$ServerConnectorManager.newConnection(ServerConnector.java:532) ~[jetty-server-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:227) ~[jetty-io-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.io.ManagedSelector.access$1500(ManagedSelector.java:60) ~[jetty-io-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.io.ManagedSelector$Accept.run(ManagedSelector.java:667) [jetty-io-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762) [jetty-util-9.4.11.v20180605.jar:9.4.11.v20180605]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680) [jetty-util-9.4.11.v20180605.jar:9.4.11.v20180605]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
47668 DEBUG (qtp764074479-22) [    ] o.e.j.u.t.QueuedThreadPool ran org.eclipse.jetty.io.ManagedSelector$Accept@35cf5a11
```

Here are the list of Security providers, which show that `OpenSSLProvider` is added. But some how `SSLEngineImpl` is created instead (by com.sun.net.ssl.internal.ssl.Provider) instead of `ConscryptEngine`
```java
ProviderList list = [sun.security.provider.Sun, sun.security.rsa.SunRsaSign, sun.security.ec.SunEC, com.sun.net.ssl.internal.ssl.Provider, com.sun.crypto.provider.SunJCE, sun.security.jgss.SunProvider, com.sun.security.sasl.Provider, org.jcp.xml.dsig.internal.dom.XMLDSigRI, sun.security.smartcardio.SunPCSC, org.conscrypt.OpenSSLProvider];
```

Any guide to solve this problem? thanks!

**Comment by sbordet** [comment]
You need to add `jetty-alpn-conscrypt-server` to the classpath.

**Comment by CaoManhDat** [comment]
Hi @sbordet, `jetty-alpn-conscrypt-server` is already present in classpath as you see an instance of conscrypt processor is created `org.eclipse.jetty.alpn.conscrypt.server.ConscryptServerALPNProcessor@2f4a2218`

**Comment by CaoManhDat** [comment]
This is quite interesting, after digging into Java security code. 
This method `SSLContext.getInstance(String protocol)` will load an instance of `SSLContextSpi`. This method will look into ProviderList. In my case 
```
ProviderList list = [sun.security.provider.Sun, sun.security.rsa.SunRsaSign, sun.security.ec.SunEC, com.sun.net.ssl.internal.ssl.Provider, com.sun.crypto.provider.SunJCE, sun.security.jgss.SunProvider, com.sun.security.sasl.Provider, org.jcp.xml.dsig.internal.dom.XMLDSigRI, sun.security.smartcardio.SunPCSC, org.conscrypt.OpenSSLProvider]
```. 

Each Provider will be ask sequentially to create an instance of `SSLContextSpi`, since `com.sun.net.ssl.internal.ssl.Provider` can provide such kind of instance, `sun.security.ssl.SSLContextImpl` is created instead of `org.conscrypt.OpenSSLContextImpl`.

Therefore by fixing this problem, this line (which is suggested by Jetty documentation)
```
Security.addProvider(new OpenSSLProvider());
```

must be changed to (therefore conscrypt provider will be asked before sun provider)
```
Security.insertProviderAt(new OpenSSLProvider(), 1);
```

**Comment by sbordet** [comment]
@CaoManhDat you should not need to call `Security.addProvider()` as this is done by `ConscryptServerALPNProcessor`.

**Comment by sbordet** [comment]
@CaoManhDat what JDK version and vendor are you using?

Because if your analysis is right, then we also would hit the same problem all the times (as the Sun provider being allocated instead of the Conscrypt one), and instead we have tests with Conscrypt that pass cleanly. :thinking:

**Comment by CaoManhDat** [comment]
I'm using OpenJDK 1.8.0_181.

**Comment by CaoManhDat** [comment]
I think what have not been test here is 

- OpenJDK
- Embedded Jetty Server with Conscrypt setup
- A http2 client connect to that server

**Comment by joakime** [comment]
@CaoManhDat did you get a solution to this issue?

**Comment by CaoManhDat** [comment]
That seems a conflict on SSL setup between Jetty client and Apache HttpClient. This issue can be closed now.

**Comment by SonaliDasgupta** [comment]
@CaoManhDat  I am facing the very same issue. Could you please let me know how you resolved it ?

**Comment by Klotzi111** [comment]
I had the same problem. And I found a solution for my case.

**My Case**
I am using Javalin (which uses jetty internally) with java 21 and the http server was working fine but the https server not.
I first thought that this is a problem to Linux (or the runtime I used in Linux) because on my Windows machine it worked.
But that was not the problem. The problem was that I used the shadowJar Gradle plugin to merge the entire classpath into a single jar file. And running the app via the bundled jar file gave me the problem. (On my Windows machine I ran the app without bundeling (gradle run) and thus it worked)
The problem was caused because the service file `META-INF/services/org.eclipse.jetty.io.ssl.ALPNProcessor$Server` collided between the `jetty-alpn-java-server` and `jetty-alpn-conscrypt-server` library. And the latter overwrote the service file of the first. And for some reason only the service implementation of `jetty-alpn-java-server` works in java 21.

**The Solution**
The fix was really simple. I had to add the following snippet to my `build.gradle` file:

```groovy
shadowJar {
  mergeServiceFiles()
}
```

This is also described on the wiki: https://imperceptiblethoughts.com/shadow/configuration/merging/#merging-service-descriptor-files

I knew that shadowJar needs to do some merging and I though those Transformers were enabled by default. But the are not.
